[Inductor] Enable Custom op Autotune Decompositions and Parameter Tuning

tianrengao opened the pull request:
This PR introduces CustomOp autotuning. It allows user to provide a CustomOpConfig:
(1) to register (optional) multiple decomposition implementations for custom operations and 
(2) to register parameter tuning knobs and values they want to tune for the decompositions
so that inductor automatically select the best-performing variant through Inductor's autotune benchmarking.

Example:
```python
 register_custom_op_autotuning(
            custom_op=my_attention_op,
            configs=[
                CustomOpConfig(attention_impl, head_dim=32, method='chunked'),
                CustomOpConfig(attention_impl, head_dim=64, method='tiled'),
                CustomOpConfig(head_dim=128), # no decompositions
            ],
            input_gen_fns={
                "query": lambda fake: torch.randn_like(fake, device='cuda'),
                "key": lambda fake: torch.randn_like(fake, device='cuda'),
                "value": lambda fake: torch.randn_like(fake, device='cuda'),
            }
    )
```

**CustomOpConfig**: Each CustomOpConfig defines exactly one autotuning variant with specific parameter values and optional decomposition implementation with PyTorch aten ops. Users can register their own tuning knobs and optional decomposition functions for the same custom operation. The system automatically benchmarks all variants to select the best performing. If no decomposition is provided in the config, the CustomOp's default implementation will be used.

**Custom Input Generation**: Users can provide custom input generators via an optional `input_gen_fns` to control how synthetic inputs are created during benchmarking. This enables more realistic performance testing by generating inputs that match expected data distributions and characteristics for each tensor argument.


**More Examples with autotune logs:**:
1. Allow user to register customOp decompositions with tuning parameters for autotuning. Example usage:
```python
from torch._inductor.kernel.custom_op import CustomOpConfig, register_custom_op_autotuning

def decompose_k_implementation(a: torch.Tensor, b: torch.Tensor, k_splits: int = 4) -> torch.Tensor:
    """Matrix multiply with k-way decomposition."""
         # Implementation...with k_splits

@torch.library.custom_op("my_lib::decompose_k", mutates_args=())
def test_decompose_k_op(
        a: torch.Tensor, b: torch.Tensor, k_splits: int
    ) -> torch.Tensor:
        return decompose_k_implementation(a, b, k_splits)

# Register autotuning with different k_splits values
register_custom_op_autotuning(
    custom_op=test_decompose_k_op,
    configs=[
        CustomOpConfig(decompose_k_implementation, k_splits=2),
        CustomOpConfig(decompose_k_implementation, k_splits=32),
        CustomOpConfig(decompose_k_implementation, k_splits=64), 
        CustomOpConfig(k_splits=128), # can make decomposition optional, then use default impl test_decompose_k_op
        CustomOpConfig(k_splits=256)
    ],
    input_gen_fns={
        "a": lambda fake: torch.randn_like(fake, device='cuda') * 0.1,
        "b": lambda fake: torch.randn_like(fake, device='cuda') * 0.1,
    }
)
```

Example result: 
```
{"num_choices": 6, "num_triton_choices": 0, "best_kernel": "test_decompose_k_autotuned_fallback_default", "best_time": 0.09980800002813339}
AUTOTUNE test_decompose_k_autotuned(256x65536, 65536x1024)
strides: [65536, 1], [1024, 1]
dtypes: torch.float16, torch.float16
  test_decompose_k_autotuned_fallback_default 0.0998 ms 100.0% 
  test_decompose_k_autotuned_decompose_k_implementation_k_splits_2_0 0.1096 ms 91.0% CustomOp decompose_k_implementation_k_splits_2
  test_decompose_k_autotuned_decompose_k_implementation_k_splits_32_1 0.1277 ms 78.2% CustomOp decompose_k_implementation_k_splits_32
  test_decompose_k_autotuned_decompose_k_implementation_k_splits_64_2 0.1454 ms 68.6% CustomOp decompose_k_implementation_k_splits_64
  test_decompose_k_autotuned_decompose_k_implementation_k_splits_128_3 0.1536 ms 65.0% CustomOp decompose_k_implementation_k_splits_128
  test_decompose_k_autotuned_decompose_k_implementation_k_splits_256_4 0.2084 ms 47.9% CustomOp decompose_k_implementation_k_splits_256
```

2. Allow user to tune parameter knob by passing the parameter and values in the CustomOpConfig. 
**Example**
```python
def mlp_variants(input_tensor, gate_weight, up_weight, down_weight, method):
    """MLP implementation with different computational approaches."""
    if method == 0:
        # Standard separate matmuls
        # ... implementation
    elif method == 1:
        # Batched approach with torch.mm
        # ... implementation
    elif method == 2:
        # Fused weights approach
        # ... implementation

@torch.library.custom_op("my_lib::mlp_op", mutates_args=())
        def mlp_op(
            input_tensor: torch.Tensor,
            gate_weight: torch.Tensor,
            up_weight: torch.Tensor,
            down_weight: torch.Tensor,
            method: int,
        ) -> torch.Tensor:
            return mlp_variants(
                input_tensor, gate_weight, up_weight, down_weight, method=method
            )

register_custom_op_autotuning(
    custom_op=mlp_op,
    configs=[
        CustomOpConfig(method=0),
        CustomOpConfig(method=1),
        CustomOpConfig(method=2),
        # method=0 is the default fallback in the original op
    ],
    input_gen_fns={
        "input_tensor": lambda fake: torch.randn_like(fake, device='cuda') * 0.1,
        "gate_weight": lambda fake: torch.randn_like(fake, device='cuda') * 0.05,
        # ... other input generators
    }
)

```

Example result: 
```
AUTOTUNE test_mlp_autotuned(4x32x512, 512x1024, 512x1024, 1024x256)
  test_mlp_autotuned_mlp_variants_method_2 0.0181 ms 100.0% CustomOp mlp_variants_method_2
  test_mlp_autotuned_mlp_variants_method_1 0.0185 ms 97.8% CustomOp mlp_variants_method_1
  test_mlp_autotuned_mlp_default_fallback_method_0 0.0198 ms 91.4% CustomOp fallback
```

### Test Suite (`test/inductor/test_custom_op_autotune.py`)

*   **RMSNorm autotuning**: Tests different RMSNorm implementations with dynamic input shapes
*   **MLP autotuning**: Tests different MLP decomposition and tuning "method" parameter
*   **DecomposeK**: Tests different k_splits values for matrix multiplication decomposition with k dim split
*   **Multi-parameter tuning**: Tests configs with multiple tuning parameters (scale_mode, chunk_size)


### Next Step:
- Enable Max-autotune with user passed in max-autotune config. https://github.com/pytorch/pytorch/pull/165526/files 
- Support inline epilogue fusion for selected best customop decomposition with surrounding elementwise ops. https://github.com/pytorch/pytorch/pull/165952/files 
- Support customop autotune considering fusion with multiTemplateBuffer. WIP

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @mlazos

tianrengao committed:
refine tests and update customop

tianrengao committed:
add tests

tianrengao committed:
refine test for rmsnorm variants and custom op, clean up code, test passed for 3 decompositions

tianrengao committed:
add test

tianrengao committed:
update test

tianrengao committed:
lint

tianrengao committed:
initial tests and modified custom op kernel template

tianrengao committed:
linter fix

tianrengao committed:
fix lint

tianrengao committed:
fix lint issues

tianrengao committed:
add fallback and seperate cpu and gpu backend

tianrengao committed:
fix lint issue and deal with device properly

eellison reviewed (commented):
good first start ! 

Would be great to consolidate a bit of the logic with existing subgraph choice.
Still needs to be handled:
- stride output logic validation
- register fn for input generation

tianrengao committed:
Tianren/custom op autotune fix (#164689)

* remove redundant catch and callable list; refine decorator to avoid using register_lowering; fix document

* wip fallback to default

* fix lint

* refine test case

* include default implemenation to the choices

* ensure test passed for default implementation correctly; fix lint

* clean up test

* refine test structure

* refine test, fix lint, remove new tempalte

* clean up code and refine

* clean up code

* simply code and lint

tianrengao committed:
revert changes and clean up code

tianrengao commented on torch/_inductor/kernel/custom_op.py:
```diff
@@ -0,0 +1,194 @@
+
+def autotune_custom_op(
+    name: str,
+    decompositions: Union[Callable[..., Any], list[Callable[..., Any]]],
```
Decompositions is always a list. Assertion added

tianrengao commented on torch/_inductor/kernel/custom_op.py:
```diff
@@ -0,0 +1,194 @@
+    """
+
+    def decorator(fn: Callable[..., Any]) -> Callable[..., Any]:
+        return register_lowering(custom_op, type_promotion_kind=None)(fn)
```
Removed register_lowering and used lowering dictionary instead in a wrapper of decorator.

tianrengao commented on test/inductor/test_custom_op_autotune.py:
```diff
@@ -0,0 +1,474 @@
+    return output
+
+
+class TestCustomOpAutoTune(TestCase):
```
Updated the test cases to cover a RMSNorm, a MLP, and a emulated DecomposeK tests . Each test contains various implementations, and asserts the numerical results are equivalent. 
The run log should show autotuning benchmark including a default implementation of the custom op.

tianrengao commented on torch/_inductor/kernel/custom_op.py:
```diff
@@ -0,0 +1,194 @@
+    Returns:
+        Optimized implementation result
+    """
+    if kwargs is None:
```
Sure, included default implementation with an ExternKernel Choice caller

tianrengao commented:
> good first start !
> 
> Would be great to consolidate a bit of the logic with existing subgraph choice. Still needs to be handled:
> 
> * stride output logic validation
> * register fn for input generation

Thanks for the review! @eellison 
I added the stride output logic validation and register fn for input generation. Also the code structure is adjusted a bit to improve readability. I also resolved comments and updated tests. Now the default implementation of customop is also used as a choice for benchmark.

tianrengao commented on torch/_inductor/select_algorithm.py:
```diff
@@ -3483,6 +3482,11 @@ def log_results(
             return
+
+        # If no timings (e.g., after fallback), skip detailed logging
+        if not timings:
```
nit: Not needed. I will disgard this change later

tianrengao committed:
remove input_gen_fns from user's interface; moving to inductor internal

tianrengao committed:
resolve input_gen_fn call and register faketensor input

tianrengao committed:
support custom op tuning for parameters; passing tuning knob as an arg

tianrengao committed:
simplify logic

youkaichao reviewed (commented):
High-level questions:

> Users can register several decomposition functions for the same custom operation

by "decomposition functions", do you mean the decomposition functions will still be traced by Dynamo/Inductor? what variants do you use to benchmark? what variants are targeted here? are they simple python code that just run as-is, or still go through Dynamo/Inductor optimizations (and at which level?)

Do you think we should give users the control over how to benchmark each variants? e.g.  for some communication ops, we might need to do a sync before measuring anything.

Does PyTorch provide autotuning cache? And what are the cache keys and cache values?

tianrengao committed:
fix cpu tests and lint

tianrengao committed:
simplify API and improve code readability

eellison commented on torch/_inductor/kernel/custom_op.py:
```diff
@@ -0,0 +1,345 @@
+
+    for key, value in kwargs.items():
+        if isinstance(value, (TensorBox, Buffer)) or (
+            hasattr(value, "dtype") and hasattr(value, "shape")
```
when are these `hasattr(value, "dtype") and hasattr(value, "shape")` necessary ? can we delete them?

eellison reviewed (commented):
another round of comments - cc @zou3519 , @BoyuanFeng please take a look

tianrengao committed:
resolve comments:fix input_gen_fns API; tensor shape infer with siz_hints instead of ir_node_to_tensor; seperated non_tensor_args for Functools.partial for kwargs ; added Single output - layout assertion; lint

tianrengao commented on torch/_inductor/codegen/subgraph.py:
```diff
@@ -207,3 +212,113 @@ def generate(  # type: ignore[override]
+                from torch.fx.experimental.proxy_tensor import make_fx
+
+                # Ensure kwargs is not None for unpacking
+                decomp_kwargs = kwargs if kwargs is not None else {}
```
The kwargs passed in SubgraphTemplate have been filtered and only contains non-tensor args. 
Processed here https://github.com/pytorch/pytorch/pull/164212/files#diff-a8f767b3bf3b50f0398a6599a9517c4a117255c5c061555883d974e233215bd0R40-R54

And being passed in from here:
https://github.com/pytorch/pytorch/pull/164212/files#diff-a8f767b3bf3b50f0398a6599a9517c4a117255c5c061555883d974e233215bd0R343

youkaichao commented:
some clarifications:

there're two concepts of "custom ops" in vLLM:

1. pytorch's custom op, registered by `direct_register_custom_op` . this is a real custom op in pytorch world, and we can call `torch.ops.vllm.xxx` to call the underlying function.
2. vLLM's custom op module, a `CustomOp` class that inherits `nn.Module`. It's mostly for dispatching the forward function to different platform implementations. Essentially it is still a `nn.Module` and is stateful.

the custom op autotune looks attractive. it should target the first kind of custom op iiuc. maybe we can augment vLLM's `direct_register_custom_op` with the proposed `register_custom_op_autotuning`. one caveat is we need to split the native torch implementation https://github.com/vllm-project/vllm/blob/44c85556216bb24ab5ccf410dbb2bcfc25c85dd5/vllm/model_executor/layers/layernorm.py#L198 into `rms_norm` and `fused_add_rms_norm`. Maybe it's not a big deal, we can ignore the `forward_native` function and just write another native `rms_norm` and `fused_add_rms_norm` implementation for inductor to compile.

> Flow: Python functions → make_fx → FX graph → Inductor compilation (no fusion, Triton kernels) → benchmark competition

It's not clear what would happen if we have some graph transforms that need to recognize the rmsnorm pattern. cc @ProExpertProg Luka knows more about this. I'm mainly concerned about how this custom op tuning interacts with the way vLLM uses inductor.

another potential usage is to tune the allreduce implementation. see https://github.com/vllm-project/vllm/blob/17838e50efc9e686fffc901dfed811b15b7d614a/vllm/distributed/device_communicators/cuda_communicator.py#L127 to check how many allreduce implementations we have. I really want to simplify them (that's why I mention distributed communication). however, I don't think people will always turn on autotune, so we need to keep all these code path anyway 😢 

btw, one more caveat: some operations use different code path for cudagraph capture and non-cudagraph capture, and that can be a problem for custom op autotune. but maybe we can leave them alone now, too complicated to consider.

eellison commented:
> It's not clear what would happen if we have some graph transforms that need to recognize the rmsnorm pattern. cc @ProExpertProg Luka knows more about this. I'm mainly concerned about how this custom op tuning interacts with the way vLLM uses inductor.

Relatedly - I'm not sure how much of the inductor pass pipeline we want to apply to these decompositions. Potentially this is something that can be opted into by the user. I think a default of no passes probably makes sense. If you're writing a decomp, this is already narrowly scoped to the point where you should be able to write it as you want it to end up..

eellison commented:
@youkaichao we've also seen instances where vllm custom ops were slower than the corresponding inductor generated decomposition. in any case, the decomposition can be optional.

zou3519 commented on test/inductor/test_custom_op_autotune.py:
```diff
@@ -0,0 +1,433 @@
+                inputs=[input_tensor, weight],
+                kwargs={"eps": eps},
+                default_impl=default_impl,
+            )
```
@tianrengao did you do anything in response to this comment? I see the PR description hasn't been updated

zou3519 commented on torch/_inductor/kernel/custom_op.py:
```diff
@@ -0,0 +1,351 @@
+            tuning_knob={
+                "head_dim": [32, 64, 128],
+                "method": ["chunked", "tiled"]
+            },
```
I feel like this API is trying to do too much.

> tuning_knob

Instead of tuning the cartesian product, it seems like it would be better to provide an explicit list of all of the configs, like how triton.autotune does it. Right now there is no way to autotune something that is not the explicit cartesian product of all the configs.

> decompositions=[standard_attention, flash_attention],
tuning_knob={

Now I'm lost -- do we autotune the cartesian product of all of the tuning_knobs and the decompositions? e.g. is we standard_attention with head_dim=32 a tuning option?

tianrengao replied:
Updated the API to be 
```        
register_custom_op_autotuning(
            torch.ops.mylib.attention.default,
            configs=[
                CustomOpConfig(attention_impl, head_dim=32, method='chunked'),
                CustomOpConfig(attention_impl, head_dim=64, method='tiled'),
                CustomOpConfig(fallback_impl),  # No params
            ],
            input_gen_fns={
                "query": lambda fake: torch.randn_like(fake, device='cuda'),
                "key": lambda fake: torch.randn_like(fake, device='cuda'),
                "value": lambda fake: torch.randn_like(fake, device='cuda'),
            }
        )
```

So it provides fine-grained control for users to combine a set of param values and decompositions.

tianrengao committed:
update API morepscificiely and use config

tianrengao committed:
param suffix and test refine

tianrengao committed:
rebase + support optional decomposition; refine customopconfig

tianrengao committed:
refactor subgraph and fix lint

tianrengao committed:
clean up code

tianrengao committed:
refine code

zou3519 commented on test/inductor/test_custom_op_autotune.py:
```diff
@@ -0,0 +1,542 @@
+        self.assertEqual(
+            compiled_result.shape, expected.shape, f"{test_name} shape mismatch"
+        )
+        torch.testing.assert_close(
```
Why not self.assertEqual? That's how we test the other stuff in PyTorch

tianrengao replied:
Sure. Will change to assertEqual later

zou3519 reviewed (approved):
Thank you, I like the new explicit api.

tianrengao committed:
resolve comments and lint. now good to land

tianrengao committed:
update var n op_overload ame and simplify logic;

tianrengao committed:
simplify tests for caffe2
